Imperative SchedulerCommand interface

timcharper opened the pull request:
Summary:
Imperative SchedulerCommand interface
We update the SchedulerLogic interace to receive LaunchCommands rather
than the use the declarative replication scheme. In the end, this is
simpler and fits the problem domain better.

We remove SpecsSnapshot concern, and clean up quite a bit of confusing
code in the process.

Docs are updated (both code and docs folder)

JIRA Issues: DCOS_OSS-5094

meln1k commented on core-models/src/main/scala/com/mesosphere/usi/core/models/SchedulerCommand.scala:
```diff
@@ -0,0 +1,38 @@
+  * Sealed trait which includes all possible events that can describe the evolution of the framework implementation's
+  * Specification state.
+  */
+sealed trait SchedulerCommand
```
👏

meln1k commented on core-models/src/main/scala/com/mesosphere/usi/core/models/StateEvent.scala:
```diff
@@ -28,12 +23,12 @@ object StateSnapshot {
 sealed trait UserError extends PodStateEvent
 
 case class PodInvalid(id: PodId, errors: Seq[String]) extends UserError
+case class PodSpecUpdated(id: PodId, newState: Option[PodSpec]) extends StateUpdated with PodStateEvent
```
It would be really cool to see some scaladoc here

meln1k reviewed (approved):
Great PR! I'd add some more docs about PodStatus class and then LGTM.

kvish commented on core-models/src/main/scala/com/mesosphere/usi/core/models/PodStatus.scala:
```diff
@@ -13,7 +13,9 @@ import org.apache.mesos.v1.{Protos => Mesos}
   * there is no specification.
   */
 case class PodStatus(id: PodId, taskStatuses: Map[TaskId, Mesos.TaskStatus]) {
+  private val activeTaskStatus = Set(Mesos.TaskState.TASK_RUNNING, Mesos.TaskState.TASK_STAGING)
```
https://github.com/apache/mesos/blob/master/include/mesos/mesos.proto#L2342

should TASK_STARTING be included here?

kvish commented on core-models/src/main/scala/com/mesosphere/usi/core/models/PodStatus.scala:
```diff
@@ -13,7 +13,9 @@ import org.apache.mesos.v1.{Protos => Mesos}
+
   def isTerminalOrUnreachable: Boolean = {
-    taskStatuses.values.forall(status => status.getState == Mesos.TaskState.TASK_RUNNING)
+    !taskStatuses.valuesIterator.forall(status => activeTaskStatus(status.getState))
```
based on above it seems like a task_starting update would be marked as terminal or unreachable

kvish commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunSpec.scala:
```diff
@@ -3,15 +3,15 @@ package com.mesosphere.usi.core.models
-  * @param shellCommand tasks' shell command
-  * @param fetch a list of artifact URIs that are passed to Mesos fetcher module and resolved at runtime
+  * @param shellCommand         tasks' shell command
+  * @param fetch                a list of artifact URIs that are passed to Mesos fetcher module and resolved at runtime
```
nit: formatting seems off here

timcharper replied:
looks like scalafmt fixed it

Commit:
Imperative SchedulerCommand interface

We update the SchedulerLogic interace to receive LaunchCommands rather
than the use the declarative replication scheme. In the end, this is
simpler and fits the problem domain better.

We remove SpecsSnapshot concern, and clean up quite a bit of confusing
code in the process.

Docs are updated (both code and docs folder)

Commit:
TASK_STARTING is also an activate state

Commit:
formatting

Commit:
Renames, docs

Commit:
Formatting again

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunningPodSpec.scala:
```diff
@@ -2,27 +2,38 @@ package com.mesosphere.usi.core.models
+  def isTerminal: Boolean
+}
+
+case class TerminalPodSpec(id: PodId) extends PodSpec {
```
Could you add docs?

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunningPodSpec.scala:
```diff
@@ -2,27 +2,38 @@ package com.mesosphere.usi.core.models
   * Framework implementation owned specification of some Pod that should be launched.
   *
-  * Pods are launched at-most-once. It is illegal to transition of [[PodSpec]] from goal terminal to goal running.
+  * Pods are launched at-most-once. It is illegal to transition of [[RunningPodSpec]] from goal terminal to goal running.
```
I'm not sure these docs still apply since there is no goal change.

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunningPodSpec.scala:
```diff
@@ -2,27 +2,38 @@ package com.mesosphere.usi.core.models
 
 import com.mesosphere.usi.core.models.resources.{ExactValue, RangeRequirement, ResourceRequirement, ScalarRequirement}
 
+sealed trait PodSpec {
```
This name is a little confusing. Is `RunSpec` implementing `PodSpec`? What do you think about naming `RunSpec` `Template`?

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/SchedulerCommand.scala:
```diff
@@ -0,0 +1,38 @@
+
+/**
+  * Sealed trait which includes all possible events that can describe the evolution of the framework implementation's
+  * Specification state.
```
Hm, these docs are user facing. I think this requires a little bit more meat. Eg a user should not care that it's sealed. What does a `Command` do? Also, is it really an event? I know this is nit-picking...

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/SchedulerCommand.scala:
```diff
@@ -0,0 +1,38 @@
+  * being launched.
+  *
+  * Submitting a launch command for a pod that is already launched (or has record of being launched) is a no-op. A
+  * unique pod id must be used for each launch.
```
That means the op is idempotent?

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -88,122 +82,37 @@ object Scheduler {
-    val (stateQueue, stateSource) = Source.queue[StateEvent](1, OverflowStrategy.backpressure).preMaterialize()
-
-    val (specQueue, specSink) = Sink.queue[SpecUpdated]().preMaterialize()
+  def asSourceAndSink(schedulerFlow: Flow[SchedulerCommand, StateOutput, NotUsed])(implicit mat: Materializer)
```
This seems like an Akka streams helper method and not a Scheduler factory method.

timcharper replied:
Agreed. Other PR

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -65,18 +61,16 @@ import scala.util.{Failure, Success}
-  type SpecInput = (SpecsSnapshot, Source[SpecUpdated, Any])
-
-  type StateOutput = (StateSnapshot, Source[StateEvent, Any])
+  type StateOutput = (StateSnapshot, Source[StateEventOrSnapshot, Any])
```
```suggestion
  type StateOutput = (StateSnapshot, Source[StateEvent, Any])
```

or simply `type StateOuput = StateEventOrSnapshot`.

timcharper replied:
Yeah, lets do this.

Commit:
Fix the tests

formatting

Commit:
mas docs

Commit:
docs

Commit:
Don't submit podSpecUpdated if same runSpec submitted twice

takirala commented on core-models/src/main/scala/com/mesosphere/usi/core/models/StateEventOrSnapshot.scala:
```diff
@@ -0,0 +1,84 @@
+  *
+  * The podSpec is cleared from USI under the following conditions:
+  *
+  * jf [[PodSpec]] is [[RunningPodSpec]], then the pod is launching and offer matching is occurring.
```
```suggestion
  * If [[PodSpec]] is [[RunningPodSpec]], then the pod is launching and offer matching is occurring.
```

takirala commented on core/src/main/scala/com/mesosphere/usi/core/logic/SpecLogic.scala:
```diff
@@ -5,60 +5,51 @@ import com.mesosphere.usi.core._
+    import com.mesosphere.usi.core.protos.ProtoConversions._
+    podStatus.taskStatuses.foldLeft(eventsBuilder) {
+      case (eventsBuilder, (taskId, status)) =>
+        eventsBuilder.withMesosCall(
```
not sure if it was intentional to shadow `eventsBuilder` here.

takirala commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunningPodSpec.scala:
```diff
@@ -3,26 +3,42 @@ package com.mesosphere.usi.core.models
+}
+
+/**
+  * Used by the scheduler to track that a pod pod should be launched (and, isn't yet.)
```
```suggestion
  * Used by the scheduler to track that a pod should be launched (and, isn't yet.)
```

takirala commented on core-models/src/main/scala/com/mesosphere/usi/core/models/StateEventOrSnapshot.scala:
```diff
@@ -0,0 +1,84 @@
+  * If [[PodSpec]] is [[TerminalPodSpec]], then USI is killing a pod and has not yet received a terminal status for the
+  * pod.
+  * @param id
+  * @param newState
```
I think it would be useful to add description for this param (as to when `newState` will be `None` vs `Some` and what it entails..)

takirala commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicHandler.scala:
```diff
@@ -2,32 +2,37 @@ package com.mesosphere.usi.core
-  * - We have built-in guarantees that evolutions to the SchedulerState can be reliably replicated by processing these
-  * events, since they led to the changes in the first place
+  * - Efficiency: Manipulating the state through events makes it efficient to know what has changed, so we can react to
+  *   change effeciently without diffing a potentially-large data structure.
```
```suggestion
  *   change efficiently without diffing a potentially-large data structure.
```

kvish commented on core-models/src/main/scala/com/mesosphere/usi/core/models/SchedulerCommand.scala:
```diff
@@ -0,0 +1,42 @@
+  *
+  * @param podId
+  */
+case class ExpungePod(podId: PodId) extends SchedulerCommand
```
Confused on the use case in which you would only want to expunge the pod. as I see it you would want to kill and expunge together

timcharper replied:
Yeah, this needs lots of thought.

jeschkies reviewed (approved):
🎉

kvish commented on core-models/src/main/scala/com/mesosphere/usi/core/models/SchedulerCommand.scala:
```diff
@@ -0,0 +1,42 @@
+
+/**
+  * Delete the pod record and any pending pod spec (running or kill) for a given pod. Does not result in kill
+  * (immediately, nor immediately).
```
wording of this comment is a little strange, did you mean (immediately, nor non-immediately)?

timcharper replied:
Yeah, reworded; great find

kvish commented on core-models/src/main/scala/com/mesosphere/usi/core/models/StateEventOrSnapshot.scala:
```diff
@@ -0,0 +1,84 @@
+  *
+  * The podSpec is cleared from USI under the following conditions:
+  *
+  * jf [[PodSpec]] is [[RunningPodSpec]], then the pod is launching and offer matching is occurring.
```
```suggestion
  * if [[PodSpec]] is [[RunningPodSpec]], then the pod is launching and offer matching is occurring.
```

timcharper replied:
Reworded

takirala reviewed (approved):
This looks awesome and I can totally see USI making fwk dev way more easier.
I think having more verbose docs in `RunningPodSpec.scala` and `SchedulerCommand.scala` would make this much more intuitive.

Commit:
Update type from StateEventOrSnapshot to StateEvent

Docs

Commit:
More docs
